utils/ablation.py

```diff
@@ -170,4 +170,29 @@ def ablate_attn_head_pos_hook(
         ablation_func = lambda x: x
     for p in pos:
         component[:, p, head_idx, :] = ablation_func(cache[f"blocks.{layer}.attn.{component_type}"][:, p, head_idx, :])
+    return component
+
+
+def ablate_resid_with_precalc_mean(
+    component: Float[Tensor, "batch ..."],
+    hook: HookPoint,
+    cached_means: Float[Tensor, "layer ..."],
+    pos_by_batch: List[List[int]],
+    layer: int = 0,
+) -> Float[Tensor, "batch ..."]:
+    """
+    Mean-ablates a batch tensor
+
+    :param component: the tensor to compute the mean over the batch dim of
+    :return: the mean over the cache component of the tensor
+    """
+    assert 'resid' in hook.name
+
+    batch_size = component.shape[0]
+    assert len(pos_by_batch) == batch_size
+
+    for i in range(batch_size):
+        for p in pos_by_batch[i]:
+            component[i, p, :] = cached_means[layer]
```

Comment:
This seems a bit inconsistent with the shapes above. I think we should use `...` in both places or neither

Reply:
done